src/class-bulk-task.php

```diff
@@ -215,6 +228,108 @@ public function run( array $args, callable $callable, string $object_type = 'wp_
 		call_user_func( [ $this, "run_${object_type}_query" ], $args, $callable );
 	}
 
+	/**
+	 * Loop through any number of rows in a CSV file efficiently with a callback, and output progress.
+	 *
+	 * @throws Exception If the CSV file does not exist or is not readable.
+	 *
+	 * @param array    $args {
+	 *     Args for the CSV query.
+	 *     @type string $csv    Path to the CSV file.
+	 *     @type int    $number Number of rows to clear cursor in each batch.
+	 * }
+	 * @param callable $callable Callback function to invoke for each row.
+	 *                           The callable will be passed a row array.
+	 */
+	public function run_csv_query( array $args, callable $callable ): void {
+
+		// Apply default arguments.
+		$args = wp_parse_args(
+			$args,
+			[
+				'csv'    => '',
+				'number' => 100,
+			],
+		);
+
+		// Ensure the CSV file exists and is readable.
+		if ( empty( $args['csv'] ) || ! is_readable( $args['csv'] ) ) {
+			throw new Exception( 'The CSV file does not exist or is not readable.' );
+		}
+
+		/**
+		 * If the CSV document was created or is read on a Legacy Macintosh computer,
+		 * help PHP detect line ending.
+		 *
+		 * @see https://php.watch/versions/8.1/auto_detect_line_endings-ini-deprecated
+		 */
+		if ( version_compare( PHP_VERSION, '8.1.0', '<' ) ) {
+			if ( ! ini_get( 'auto_detect_line_endings' ) ) {
+				ini_set( 'auto_detect_line_endings', '1' );
+			}
+		}
+
+		// It assumes that the file is encoded in UTF-8.
+		try {
+			$csv = new SplFileObject( $args['csv'], 'r' );
+		} catch ( Exception $e ) {
+			throw new Exception( $e->getMessage() ); // phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
+		}
+
+		// Set the CSV flags.
+		$csv->setFlags( SplFileObject::READ_CSV | SplFileObject::READ_AHEAD | SplFileObject::SKIP_EMPTY | SplFileObject::DROP_NEW_LINE );
+
+		// Set the min ID from the cursor.
+		$this->min_id = $this->cursor->get();
+
+		// Set the max ID from CSV file.
+		$csv->seek( PHP_INT_MAX );
+		$this->max_id = $csv->key();
+
+		// Turn off some automatic behavior that would slow down the process.
+		$this->before_run();
+
+		/**
+		 * We do not run rows in batches since the CSV file outputs all rows at once.
+		 *
+		 * But we need it for the cursor, progress to be updated, and resetting
+		 * the object cache.
+		 */
+		$batch_size = 0;
+
+		// All systems go.
+		foreach ( $csv as $row ) {
+			$line_number = $csv->key();
+
+			// Skip lines outside of the range.
+			if ( $line_number < $this->min_id || $line_number > $this->max_id ) {
```

Comment:
Is this possible?  If so, should it display a warning just incase there's an issue and some lines are getting skipped?

Reply:
The first conditional it is! When we use the cursor, we don't want to run the callback on the rows already run.

```
$line_number < $this->min_id 
current line < the last line saved in the cursor
99 < 100 <- don't run on rows below 100.
```

The other conditional, `$line_number > $this->max_id` is unlikely to happen since max_id equates to the last row number. But added just in case.